python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,618 @@
+
+    def __init__(self, repr_string: str, value):
+        self._repr_string = repr_string
+        if value is None:
+            raise TypeError("Cannot set value of BaseLiteral to None")
+        self._value = value
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    @property
+    def value(self):
+        return self._value
+
+    def __repr__(self):
+        return self._repr_string
+
+    def __str__(self):
+        return str(self._value)
+
+
+@singledispatch
+def literal(value: Any) -> BaseLiteral:
+    """
+    A generic Literal factory to construct an iceberg Literal based on python primitive data type
+    using dynamic overloading
+
+    Args:
+        value(python primitive type): the value to be associated with literal
+
+    Example:
+        from iceberg.expressions.literals import literal
+        >>> literal(123)
+        <IntegerLiteral(123)>v
+    """
+    raise TypeError(f"Unimplemented Type Literal for value: {value}")
+
+
+@literal.register
+def _(value: bool):
+    return BooleanLiteral(value)
+
+
+@literal.register
+def _(value: int):
+    """
+    Upgrade to long if python int is outside the JAVA_MIN_INT and JAVA_MAX_INT
+    """
+    if value < IntegerType.min or value > IntegerType.max:
+        return LongLiteral(value)
+    return IntegerLiteral(value)
+
+
+@literal.register
+def _(value: float):
+    """
+    Upgrade to double if python float is outside the JAVA_MIN_FLOAT and JAVA_MAX_FLOAT
+    """
+    if value < FloatType.min or value > FloatType.max:
+        return DoubleLiteral(value)
+    return FloatLiteral(value)
+
+
+@literal.register
+def _(value: str):
+    return StringLiteral(value)
+
+
+@literal.register
+def _(value: uuid.UUID):
+    return UUIDLiteral(value)
+
+
+@literal.register
+def _(value: bytes):
+    return FixedLiteral(value)
+
+
+@literal.register
+def _(value: bytearray):
+    return BinaryLiteral(value)
+
+
+@literal.register
+def _(value: Decimal):
+    return DecimalLiteral(value)
+
+
+class ComparableLiteral(BaseLiteral):
+    def __init__(self, repr_string: str, value):
+        super(ComparableLiteral, self).__init__(repr_string, value)
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        return self.value == other.value
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __lt__(self, other):
+        return self.value < other.value
+
+    def __gt__(self, other):
+        return self.value > other.value
+
+    def __le__(self, other):
+        return self.value <= other.value
+
+    def __ge__(self, other):
+        return self.value >= other.value
+
+
+class AboveMax(Literal, Singleton):
+    def value(self):
+        raise RuntimeError("AboveMax has no value")
+
+    def to(self, type_var):
+        raise RuntimeError("Cannot change the type of AboveMax")
+
+    def __str__(self):
+        return "AboveMax"
```

Comment:
https://github.com/apache/iceberg/pull/4262#discussion_r829375822 We could also generalize this to the Literal:
```
➜  python git:(master) python3
Python 3.9.10 (main, Jan 15 2022, 11:48:04) 
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from iceberg.types import Singleton
>>> 
>>> class Literal:
...     def to(self, type_var):
...         raise NotImplementedError()
...     def __str__(self):
...         return type(self).__name__
... 
>>> class AboveMax(Literal, Singleton):
...     def value(self):
...         raise RuntimeError("AboveMax has no value")
...     def to(self, type_var):
...         raise RuntimeError("Cannot change the type of AboveMax")
... 
>>> str(AboveMax())
'AboveMax'
```

Reply:
Thank you @Fokko , updated to save define `__str__` for both AboveMax and BelowMin